test_util/src/pty.rs

```diff
@@ -0,0 +1,437 @@
+use std::collections::HashMap;
+use std::io::Read;
+use std::path::Path;
+
+pub trait Pty: Read {
+  fn write_text(&mut self, text: &str);
+
+  fn write_line(&mut self, text: &str) {
+    self.write_text(&format!("{}\n", text));
+  }
+
+  /// Reads the output to the EOF.
+  fn read_all_output(&mut self) -> String {
+    let mut text = String::new();
+    self.read_to_string(&mut text).unwrap();
+    text
+  }
+}
+
+#[cfg(unix)]
+pub fn create_pty(
+  program: impl AsRef<Path>,
+  args: &[&str],
+  cwd: impl AsRef<Path>,
+  env_vars: Option<HashMap<String, String>>,
+) -> Box<dyn Pty> {
+  let fork = pty::fork::Fork::from_ptmx().unwrap();
+  if fork.is_parent().is_ok() {
+    Box::new(unix::UnixPty { fork })
+  } else {
+    std::process::Command::new(program.as_ref())
+      .current_dir(cwd)
+      .args(args)
+      .envs(env_vars.unwrap_or_default())
+      .spawn()
+      .unwrap()
+      .wait()
+      .unwrap();
+    unreachable!();
+  }
+}
+
+#[cfg(unix)]
+mod unix {
+  use std::io::Read;
+  use std::io::Write;
+
+  use super::Pty;
+
+  pub struct UnixPty {
+    pub fork: pty::fork::Fork,
+  }
+
+  impl Drop for UnixPty {
+    fn drop(&mut self) {
+      self.fork.wait().unwrap();
+    }
+  }
+
+  impl Pty for UnixPty {
+    fn write_text(&mut self, text: &str) {
+      let mut master = self.fork.is_parent().unwrap();
+      master.write_all(text.as_bytes()).unwrap();
+    }
+  }
+
+  impl Read for UnixPty {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+      let mut master = self.fork.is_parent().unwrap();
+      master.read(buf)
+    }
+  }
+}
+
+#[cfg(target_os = "windows")]
+pub fn create_pty(
+  program: impl AsRef<Path>,
+  args: &[&str],
+  cwd: impl AsRef<Path>,
+  env_vars: Option<HashMap<String, String>>,
+) -> Box<dyn Pty> {
+  let pty = windows::WinPseudoConsole::new(
+    program,
+    args,
+    &cwd.as_ref().to_string_lossy().to_string(),
+    env_vars,
+  );
+  Box::new(pty)
+}
+
+#[cfg(target_os = "windows")]
+mod windows {
```

Comment:
After I wrote most of this implementation, Ryan pointed me in the direction of using [portable_pty](https://docs.rs/portable-pty/0.4.0/portable_pty/). For some reason, it did not build for me out of the box and I did not see a way for it to avoid deadlocking on exit without changing the Pty trait to have the caller need to manage creating a new thread to read the output (see below how I inverted this so the thread is hidden in the implementation), so I just kept what I have here. I think having this own implementation is better now that the work is done, but had I done it over again I probably would have done a way to reduce the effort here. That said, we'll be able to easily and more quickly improve this now that it's here.